core/domain/classifier_services.py

```diff
@@ -576,6 +578,8 @@ def migrate_state_training_jobs(
         exp_id, version=exp_version)
     interaction_id = exploration.states[state_name].interaction.id
 
+    # Ruling out the possibility of None for mypy type checking.
```

Comment:
Similarly in line 584, we are using interaction_id to fetch a value from `feconf.INTERACTION_CLASSIFIER_MAPPING` and in `feconf.INTERACTION_CLASSIFIER_MAPPING` there is no key which is defined as None.

So to handle this I used assertion here!

Reply:
Please see above.

Reply:
Similar situation as above, no validation exists for interaction.id in callers

Reply:
Finally found something!

This method is only used in controllers' classifier.py and there we are already validating the `interaction_id`.

https://github.com/oppia/oppia/blob/5c6c538ec98b9de3f3d95c1222d66bd2a31bda2a/core/controllers/classifier.py#L146

Reply:
Good. Then you can be sure that, with the existing code, interaction_id can't be None. So add an "if interaction_id is None" exception and update the docstring to explain the validity conditions for the input. Then, you can add an assertion after that if it is still necessary.

The assertion on its own is not sufficient here because someone may write another function that calls this one in the future. In general I think it is a good rule of thumb to use assertions only for conditions that have been already validated within the same function. Please add this to the wiki page (along with an explanation) if it isn't there already.

Reply:
If clause is added and docstring is also updated!
Also, a PR to update the backend annotations doc is raised PTAL!. https://github.com/oppia/oppia-web-developer-docs/pull/16

Reply:
> use assertions only for conditions that have been already validated within the same function.

If the conditions are already validated within the same function then I doubt that we really need any assertion.

> The assertion on its own is not sufficient here because someone may write another function that calls this one in the future.

If someone calls this function with Optional type and we have assertion written in the called function, then assertion will caught this situation and throw an assertion error that value cannot be None.